src/model_query.cc

```diff
@@ -0,0 +1,54 @@
+/*!
+ * Copyright (c) 2024 by Contributors
+ * \file model_query.cc
+ * \author Hyunsu Cho
+ * \brief Methods for querying various properties of tree models
+ */
+#include <cstdint>
+#include <queue>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include <treelite/tree.h>
+
+namespace {
+
+template <typename ThresholdType, typename LeafOutputType>
+std::uint32_t GetDepth(treelite::Tree<ThresholdType, LeafOutputType> const& tree) {
+  // Visit all trees nodes in breath-first order
```

Comment:
Not really, I chose breadth-first traversal because I was the most comfortable with it.

> I just keep track of the maximum depth in a depth-first traversal, which keeps us from having to track the depth of each node separately.

Not quite sure about this part. Don't you still have to keep track of the depth of each node somewhere, either in the call stack (if using recursive function calls) or in a `std::stack<std::pair<int, int>>` (if using an explicit stack) ?

Reply:
Probably easier to show this in code, so here's a [gist](https://gist.github.com/wphicks/9575e83d31ba1feeca8a55c7c9071603). Basically, you can just track the depth and max depth, not the depth of each node separately.